microcloud/cmd/microcloud/main_init.go

```diff
@@ -272,25 +272,26 @@ func Bootstrap(sh *service.ServiceHandler, peers map[string]string) error {
 
 		return nil
 	})
-	if err != nil {
-		return err
-	}
+}
 
-	tokens := map[service.ServiceType][]string{}
+func AddPeers(sh *service.ServiceHandler, peers map[string]string, localDisks map[string][]lxdAPI.ClusterMemberConfigKey) error {
+	joinConfig := map[service.ServiceType][]mdns.JoinConfig{}
 	for serviceType := range sh.Services {
-		tokens[serviceType] = make([]string, len(peers))
+		joinConfig[serviceType] = make([]mdns.JoinConfig, len(peers))
```

Comment:
I would encourage avoiding making slices a specific size, but instead make them a specific capacity (if you know it).

So here would be `joinConfig[serviceType] = make([]mdns.JoinConfig, 0, len(peers))` and then you can populate it below using `append()`. That way the code becomes less brittle if in the future an condition logic is added that means that not all items in the loop are added to the slice, as this would end up leaving zero-intialised empty item(s) at the end of the slice.

It also avoids initialising the slice with entities that are then going to get replaced immediately afterward in loop.

Reply:
This is actually gone from the final commit, but to clarify a bit about my intentions here, I actually deliberately made this a specific sized slice because this slice was populated in an asynchronous function. I believe an append would cause some race issues.

Reply:
Interesting. By async do you mean concurrently from multiple go routines? If yes, then yes appending concurrently would certainly be an issue. Do you have synchronisation to ensure that map isn't read until all possible updates have finished?

A comment here (and in similar places) would be useful to explain why this is so, as its not clear this would be called concurrently. Thanks

Reply:
Ah yes I see now that sh.RunAsync is actually running the function concurrently for all services.
Perhaps `RunAsync` could be renamed to clarify its actually `RunConcurrent`? (as async doesn't imply concurrency).

Reply:
Good point, yeah I often conflate the two but I shouldn't.